File: src/@Genetics/Task.cpp
#include <stdexcept>
#include <iomanip>

#include "Task.h"
#include "Processor.h"

void Task::assign_processor(const Processor *processor)
{
	if (processor->type_count < type + 1)
		throw std::runtime_error("The processor does not have such type.");

	this->processor = processor;

	/* t = NC / f */
	duration = processor->nc[type] / processor->frequency;

	/* P = Ceff * f * V^2 */
	power = processor->ceff[type] * processor->frequency *
		processor->voltage * processor->voltage;
}

void Task::propagate_start(double time)
{
	/* We might already have an assigned start time with a larger value */
	if (!(start < time)) return;

	start = time;
	time = time + duration;

	/* Shift data dependent tasks */
	size_t size = children.size();
	for (size_t i = 0; i < size; i++)
		children[i]->propagate_start(time);

	/* Shift space dependent tasks (the same core) */
	if (successor) successor->propagate_start(time);
}

void Task::propagate_asap(double time)
{
	/* We might already have an assigned ASAP time with a larger value */
	if (!(asap < time)) return;

	asap = time;
	time = time + duration;

	/* Shift data dependent tasks */
	size_t size = children.size();
	for (size_t i = 0; i < size; i++)
		children[i]->propagate_asap(time);
}

void Task::propagate_alap(double time)
{
	time = std::max(0.0, time - duration);

	/* We might already have an assigned ALAP time with a smaller value */
	if (!(time < alap)) return;

	alap = time;
	mobility = std::max(0.0, alap - asap);

	/* Shift data dependent tasks */
	size_t size = parents.size();
	for (size_t i = 0; i < size; i++)
		parents[i]->propagate_alap(time);
}

std::ostream &operator<< (std::ostream &o, const Task *task)
{
	o.precision(2);
	o.flags(std::ios::fixed);

	o	<< std::setw(4) << task->id << " ( "
		<< std::setw(4) << (task->processor ? task->processor->id : -1) << " : "
		<< std::setw(4) << task->type << " : "
		<< std::setw(8) << task->start << " : "
		<< std::setw(8) << task->duration << " : "
		<< std::setw(8) << task->asap << " : "
		<< std::setw(8) << task->mobility << " : "
		<< std::setw(8) << task->alap << " ) -> [ ";

	size_t children_count = task->children.size();
	for (size_t i = 0; i < children_count; i++) {
		o << task->children[i]->id;
		if (i + 1 < children_count) o << ", ";
	}

	o << " ] " << std::endl;

	return o;
}

File: msrc/+TestCase/Task.m
classdef Task < handle
  properties (SetAccess = private)
    % General
    id
    name
    type

    % Timing
    duration
    start     % Actual start time (mapped and scheduled)
    asap      % ASAP, as soon as possible
    alap      % ALAP, as late as possible
    mobility  % ALAP - ASAP

    % Dependencies in the graph
    isLeaf
    isRoot
    parents
    children

    % Dependencies on the same core
    ancestor
    successor
  end

  methods
    function task = Task(id, name, type)
      % General
      task.id = id;
      task.name = name;
      task.type = type;

      % Timing
      task.duration = 0;
      task.start = -Inf;
      task.asap = -Inf;
      task.alap = Inf;
      task.mobility = 0;

      % Dependencies in the graph
      task.isLeaf = true;
      task.isRoot = true;
      task.parents = {};
      task.children = {};

      % Dependencies on the same core
      task.ancestor = [];
      task.successor = [];
    end

    function resetMapping(task)
      task.start = -Inf;
      task.ancestor = [];
      task.successor = [];
    end

    function scale(task, factor)
      task.start = task.start * factor;
      task.duration = task.duration * factor;
    end

    function addParent(task, parent)
      task.parents{end + 1} = parent;
      task.isRoot = false;
    end

    function addChild(task, child)
      task.children{end + 1} = child;
      task.isLeaf = false;
    end

    function setAncestor(task, ancestor)
      task.ancestor = ancestor;
    end

    function setSuccessor(task, successor)
      task.successor = successor;
    end

    function assignDuration(task, time)
      task.duration = time;
    end

    function propagateStartTime(task, time)
      % We might already have an assigned start time with larger value
      if ~(task.start < time), return; end

      task.start = time;
      time = time + task.duration;

      % Shift data dependent tasks
      for child = task.children
        child{1}.propagateStartTime(time);
      end

      % Shift space dependent tasks (the same core)
      if ~isempty(task.successor)
        task.successor.propagateStartTime(time);
      end
    end

    function propagateASAP(task, asap)
      % We might already have an assigned ASAP with larger value
      if ~(task.asap < asap), return; end

      task.asap = asap;
      asap = asap + task.duration;

      % Shift data dependent tasks
      for child = task.children
        child{1}.propagateASAP(asap);
      end
    end

    function propagateALAP(task, time)
      % As late as possible
      alap = max(0, time - task.duration);

      % We might already have an assigned ALAP with smaller value
      if ~(alap < task.alap), return; end

      task.alap = alap;

      % Mobility = ALAP - ASAP
      task.mobility = max(0, alap - task.asap);

      % Shift data dependent tasks
      for parent = task.parents
        parent{1}.propagateALAP(alap);
      end
    end
  end
end
